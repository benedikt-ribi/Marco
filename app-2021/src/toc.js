export class tocBuilder {
  applyToc(minLevel = 1, maxLevel = 6) {
    const allLevels = [...Array(7).keys()];
    const levels = allLevels.slice(minLevel, maxLevel + 1);
    const headingLevels = levels.map(l => `h${l}`);
    const headingSelectors = headingLevels.join();

    const headings = document.querySelectorAll(headingSelectors);

    let toc = "<ul>";
    /** @type Element */
    let currentLevel = undefined;

    const getLevel = n => n[1];

    for (const heading of headings) {
      const headingLevel = getLevel(heading.tagName);

      if (currentLevel) {
        if (currentLevel < headingLevel) {
          toc += "<ul>\n";
        } else if (currentLevel > headingLevel) {
          for (let index = 0; index < currentLevel; index++) {
            toc += "</ul>\n";
          }
          currentLevel = headingLevel;
          toc += "<ul>";
        }
      }

      heading.id = heading.textContent.replaceAll(/[^A-Zöäü0-9]/gi, "-");

      toc += `<li><a href="#${heading.id}">${heading.textContent}</a></li>\n`;

      currentLevel = headingLevel;
    }

    toc += "</ul>";

    const tocElement = document.querySelector("#toc");
    if (tocElement) {
      tocElement.innerHTML = toc;
    }
  }
}
