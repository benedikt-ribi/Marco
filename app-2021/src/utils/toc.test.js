/**
 * @jest-environment jsdom
 */

import { describe, expect, test } from "@jest/globals";
import { tocBuilder } from "./toc";

describe("TOC Builder", () => {
  test("build ", () => {
    document.body.innerHTML = `
      <h1>1</h1>
      <h2>1.1</h2>
      <h3>1.1.1</h3>
      <h3>1.1.2</h3>
      <h2>1.2</h2>
      <h2>1.3</h2>
      <h3>1.3.1</h3>
      <h4>1.3.1.1</h4>
      <h4>1.3.1.2</h4>
      <h3>1.3.2</h3>
      <nav id="toc"></nav>
    `;

    new tocBuilder().applyToc(2, 4);

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
