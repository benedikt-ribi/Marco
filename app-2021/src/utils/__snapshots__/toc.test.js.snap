// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TOC Builder build  1`] = `
"
      <h1>1</h1>
      <h2 id=\\"1-1\\">1.1</h2>
      <h3 id=\\"1-1-1\\">1.1.1</h3>
      <h3 id=\\"1-1-2\\">1.1.2</h3>
      <h2 id=\\"1-2\\">1.2</h2>
      <h2 id=\\"1-3\\">1.3</h2>
      <h3 id=\\"1-3-1\\">1.3.1</h3>
      <h4 id=\\"1-3-1-1\\">1.3.1.1</h4>
      <h4 id=\\"1-3-1-2\\">1.3.1.2</h4>
      <h3 id=\\"1-3-2\\">1.3.2</h3>
      <nav id=\\"toc\\"><ul><li><a href=\\"#1-1\\">1.1</a></li>
<ul>
<li><a href=\\"#1-1-1\\">1.1.1</a></li>
<li><a href=\\"#1-1-2\\">1.1.2</a></li>
</ul>
</ul>

<ul><li><a href=\\"#1-2\\">1.2</a></li>
<li><a href=\\"#1-3\\">1.3</a></li>
<ul>
<li><a href=\\"#1-3-1\\">1.3.1</a></li>
<ul>
<li><a href=\\"#1-3-1-1\\">1.3.1.1</a></li>
<li><a href=\\"#1-3-1-2\\">1.3.1.2</a></li>
</ul>
</ul>
</ul>

<ul><li><a href=\\"#1-3-2\\">1.3.2</a></li>
</ul></nav>
    "
`;
